### PR TITLE
There are still naming collisions for Windows feature functions

### PR DESCRIPTION
**Pull request by splatteredbits**
Fixes issue #55 .

### CODE REVIEW

File: Carbon/Functions/Get-Group.ps1 [hunk 1]
```diff
@@ -38,24 +38,54 @@ function Get-CGroup
...
+                    }
+                Write-Timing ('                           FindAll()  End')
+                break
+            }
+            catch
+            {
+                Write-Timing ('                           FindAll()  Failed')
+                $_ | Out-String | Write-Debug 
+                
+                if( $lastTry )
```
Comment (DecoyJoe) [inline Carbon/Functions/Get-Group.ps1]: This isn't set anywhere.

File: Carbon/Functions/Get-Group.ps1 [hunk 2]
```diff
@@ -38,24 +38,54 @@ function Get-CGroup
...
                 {
-                    return $_.Name -eq $Name
+                    Write-Error ('We''ve tried {0} times to read groups, but keep getting exceptions. We''re giving up. Here''s the last exception we got: {1}' -f $maxTries,$_) -ErrorAction $ErrorActionPreference
+                    return
                 }
-                return $true
-            } |
-            Tee-Object -Variable 'groups'
+
+                $numErrors = $Global:Error.Count - $numErrorsBefore
```
Comment (DecoyJoe) [inline Carbon/Functions/Get-Group.ps1]: `$numErrorsBefore` isn't set anywhere.

File: Test/Set-IisWebsiteID.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    
+    It 'should set same ID on same website' {
+        $Error.Clear()
+        $currentSite = Get-IisWebsite -SiteName $siteName
+        $currentSite.Start()
+        Set-IisWebsiteID -SiteName $siteName -ID $currentSite.ID -ErrorAction SilentlyContinue
+        $Error.Count | Should Be 0
+        $updatedSite = Get-IisWebsite -SiteName $siteName
+        $updatedSite.ID | Should Be $currentSite.ID
+        $updatedSite.State | Should Be 'Started'
```
Comment (DecoyJoe) [inline Test/Set-IisWebsiteID.Tests.ps1]: Looks like it's all Pester 3 syntax.

File: Carbon/Functions/Uninstall-WindowsFeature.ps1 [hunk 4]
```diff
@@ -1,6 +1,6 @@
 
 # This function should only be available if the Windows PowerShell v3.0 Server Manager cmdlets aren't already installed.
-if( -not (Get-Command -Name 'Get-CWindowsFeature*' | Where-Object { $_.ModuleName -ne 'Carbon' }) )
+if( -not (Get-Command -Name 'Get-WindowsFeature*' | Where-Object { $_.ModuleName -ne 'Carbon' }) )
```
Comment (DecoyJoe) [inline Carbon/Functions/Uninstall-WindowsFeature.ps1]: Did you mean to check for `Get-WindowsFeature` or `Uninstall-WindowsFeature`?

File: Carbon/Functions/Install-WindowsFeature.ps1 [hunk 5]
```diff
@@ -1,6 +1,6 @@
 
 # This function should only be available if the Windows PowerShell v3.0 Server Manager cmdlets aren't already installed.
-if( -not (Get-Command -Name 'Get-CWindowsFeature*' | Where-Object { $_.ModuleName -ne 'Carbon' }) )
+if( -not (Get-Command -Name 'Get-WindowsFeature*' | Where-Object { $_.ModuleName -ne 'Carbon' }) )
```
Comment (DecoyJoe) [inline Carbon/Functions/Install-WindowsFeature.ps1]: Did you mean to check for `Get-WindowsFeature` or `Install-WindowsFeature`?

File: Carbon/Functions/Get-User.ps1 [hunk 6]
```diff
@@ -58,17 +55,47 @@ function Get-CUser
...
+            }
+            catch
+            {
+                Write-Timing ('                         FindAll()  Failed')
+                $_ | Out-String | Write-Debug 
+
+                if( $lastTry )
                 {
-                    return $_.SamAccountName -eq $UserName
+                    Write-Error -Message ('We tried {0} times to read user information, but kept getting exceptions. We''ve given up. Here''s the last error we got: {1}.' -f $maxTries,$_)
```
Comment (DecoyJoe) [inline Carbon/Functions/Get-User.ps1]: Shouldn't this include `-ErrorAction $ErrorActionPreference`?

File: Carbon/Functions/Uninstall-WindowsFeature.ps1 (hunk 4 above)
Comment (splatteredbits) [reply Carbon/Functions/Uninstall-WindowsFeature.ps1]: What we're really checking for is the presence of the ServerManager module. That's why the check is always for `Get-WindowsFeature`. By design.

File: Carbon/Functions/Install-WindowsFeature.ps1 (hunk 5 above)
Comment (splatteredbits) [reply Carbon/Functions/Install-WindowsFeature.ps1]: By design.

File: Carbon/Functions/Get-Principal.ps1 [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                    Where-Object -FilterScript $Filter
+                Write-Timing ('                                FindAll()  End')
+                break
+            }
+            catch
+            {
+                Write-Timing ('                                FindAll()  Failed')
+                $_ | Out-String | Write-Debug 
+
+                $lastTry = $tryNum -ge $maxTries
```
Comment (DecoyJoe) [inline Carbon/Functions/Get-Principal.ps1]: Will this evaluate to true? It looks like the `while` loop's last iteration will be `99` so this won't evaluate to true, e.g. `99 -ge 100`. I think you need to change the operator in the `while` condition to `-le`?

Comment (splatteredbits) [reply Carbon/Functions/Get-Principal.ps1]: It starts at `0`. After the `while` condition first evaluates, it's `1`. The `++` postfix operator increments the value _after_ comparing it to `$maxTries`. Inside the loop, on the last iteration, the value will be `100` not `99`.

Comment (DecoyJoe) [reply Carbon/Functions/Get-Principal.ps1]: Perfect, wasn't sure if the compare happened before or after the increment.

**Pull request merged**